### PR TITLE
Remove coqide ml4

### PR DESCRIPTION
**Pull request by SkySkimmer**
Also remove ml4 file handling from the make based build system (coq_makefile still has it ofc, and dune because I'm not sure about removing it there).

### CODE REVIEW

**Comment by SkySkimmer** [comment]
The windows log is too verbose, I can't tell if the failure is spurious.

**Comment by gares** [comment]
I've mixed feelings:
- I'm happy we are so close to not needing camlp5
- ifdefs are ugly but extremely useful
- the solution based on sed is frankly worse than the status quo

What I think would be reasonable is to add basic ifdef support to coqpp ( @ppedrot ).
If this is too hard, then I propose to generate the relevant piece of ml at configure time (or its equivalent for dune) by providing files such as `foo.ml.osx` and copying it to `foo.ml`.

**Comment by SkySkimmer** [comment]
I don't think this quite fits coqpp.
Here's an alternative using m4 instead of custom sed stuff https://github.com/SkySkimmer/coq/commit/07de4a2c99fc3866119f61a88d7e8fc5c94be64f

> If this is too hard, or not relevant enough to others, then I propose to generate the relevant piece of ml at configure time (or its equivalent for dune) by providing files such as foo.ml.osx and copying it to foo.ml.

Not sure how to do such things in dune.

**Comment by gares** [comment]
> Not sure how to do such things in dune.

We do have dune "rules" that copy files around already.
```
./checker/dune:(rule (copy %{project_root}/kernel/names.ml names.ml))
```
I did not test it but something like
```
(rule (copy %{project_root}/ide/foo.ml.%{ocaml-config:os_type} foo.ml))
```
could work

**Comment by gares** [comment]
> Here's an alternative using m4

O_o

**Comment by SkySkimmer** [comment]
When I searched for language agnostic preprocessors it's what came out.

**Comment by Zimmi48** [comment]
@gares Emilio argued strongly about adding anything to configure. There is no equivalent of configure in Dune, only equivalents of steps in the Makefile. But this doesn't change anything right? You are just proposing to add a rule copying the right file instead of emulating `IFDEF`. It doesn't matter if this is done in Makefile or in configure.

**Comment by gares** [comment]
>  You are just proposing to add a rule copying the right file instead of emulating `IFDEF`. It doesn't matter if this is done in Makefile or in configure.

Exactly. It should be trivial in both Make and Dune, and should be cleaner than sed or m4.

**Review by ejgallego — COMMENTED** [review]
Indeed it is a bit unfortunate that we don't have a preprocessor in coq, that could be very useful [but also open a large can of worms]

I am fine with the m4 solution, I will make Dune do the right thing in other archs. 

But I am also wondering whether we should vendor cppo. As far as I understand that should be a quite clean solution [specially when we are Dune-only]

**Comment by Zimmi48** [comment]
The question of it being at configure time or at make time is irrelevant to this discussion.
Both @ppedrot and @gares support a simple `cp`-based solution so I think this is the one which should get implemented.
Vendoring a pre-processor just for one file seems completely overkill.

**Review by ppedrot — CHANGES_REQUESTED** [review]
Yep, please implement a cp-based mechanism instead.

**Review by gares — CHANGES_REQUESTED** [review]
please use `cp`

**Comment by SkySkimmer** [comment]
Here's the cp version.

**Review by ppedrot — APPROVED** [review]

File: Makefile.build [hunk 1]
```diff
@@ -259,6 +259,7 @@ CAMLP5DEPS:=grammar/grammar.cma
 CAMLP5USE=pa_extend.cmo q_MLast.cmo pa_macro.cmo -D$(CAMLVERSION)
 
 PR_O := $(if $(READABLE_ML4),pr_o.cmo,pr_dump.cmo)
+# XXX unused but should be used for mlp files
```
Comment (gares) [inline Makefile.build]: Can you remind me what mlp files are? Weren't they used just for camlp4 <-> camlp5 compatibility?

**Review by gares — APPROVED** [review]
LGTM

I let @ejgallego comment on dune. The code I proposed used `%{ocaml-config:os_type}` that is like `Sys.os_type` eg `Unix` or `Windows` ... here we use `X11` or `QUARTZ` or `WIN32`... so there is a mapping to be done that could be avoided by using the values `os_type` holds.

**Comment by SkySkimmer** [comment]
CI is reliably failing on lib/hook.cmo but I can't reproduce locally. What is going on?

File: Makefile.build (hunk 1 above)
Comment (SkySkimmer) [reply Makefile.build]: camlp4 files that don't depend on VERNAC EXTEND &co (ie the ones used to define it)

**Comment by gares** [comment]
did you try with -j1 on CI?

**Comment by SkySkimmer** [comment]
OK it works now, the issue was a bad cherry pick of the second commit from the preprocessor branch.

**Comment by ppedrot** [comment]
I'm assigning this PR and will merge promptly if nobody objects.

**Pull request merged**